src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/CoreCLR/WebRequestPSCmdlet.CoreClr.cs

```diff
@@ -174,6 +177,50 @@ internal virtual HttpClient GetHttpClient(bool handleRedirect)
                 handler.ServerCertificateCustomValidationCallback = HttpClientHandler.DangerousAcceptAnyServerCertificateValidator;
                 handler.ClientCertificateOptions = ClientCertificateOption.Manual;
             }
+            else if (CertificateValidationScript != null)
+            {
+                // validationCallBackWrapper wraps the certificateValidationClosure ScriptBlock so the async callback can set the default
+                // PowerShell Runspace in the async thread to the Runspace of the current thread. This provides the ScriptBlock access to 
+                // the current scope.
+                Runspace currentRunspace = Runspace.DefaultRunspace;
+                ScriptBlock certificateValidationClosure = CertificateValidationScript.GetNewClosure();
+                Func<HttpRequestMessage,X509Certificate2,X509Chain,SslPolicyErrors,bool> validationCallBackWrapper = 
+                    delegate(HttpRequestMessage httpRequestMessage, X509Certificate2 x509Certificate2, X509Chain x509Chain, SslPolicyErrors sslPolicyErrors)
+                    {
+                        Runspace previousRunspace = Runspace.DefaultRunspace;
+                        Runspace.DefaultRunspace = currentRunspace;
+
+                        // Set Script scope automatic variables for the closure.
+                        PSVariableIntrinsics closureVi = certificateValidationClosure.Module.SessionState.PSVariable;
+                        closureVi.Set("HttpRequestMessage", httpRequestMessage);
+                        closureVi.Set("X509Certificate2", x509Certificate2);
+                        closureVi.Set("X509Chain", x509Chain);
+                        closureVi.Set("SslPolicyErrors", sslPolicyErrors);
+                        closureVi.Set("ErrorActionPreference", "Stop");
```

Comment:
Maybe put `x509Certificate2`on first position because the callback is for it?
Another thought is to pass a PSObject as $PSItem with the properties (x509Certificate2, httpRequestMessage, ...).

Reply:
All 4 are equally important for validating the certificate. 

```powershell
$Script = {
    Return (
        $HttpRequestMessage.RequestUri.AbsoluteUri -eq 'https://www.google.com/' -and
        $X509Certificate2.Subject -eq 'CN=www.google.com, O=Google Inc, L=Mountain View, S=California, C=US' -and
        $X509Chain.ChainElements[2].Certificate.Thumbprint -eq 'DE28F4A4FFE5B92FA3C503D1A349A7F9962A8212' -and
        $SslPolicyErrors -eq 'None'
    )
}
```

I wanted to mirror the ,NET callback as much as possible to prevent confusion.

To be clear, `$args[0]`, `$X509Certificate2`, and `Param( [Parameter(Position=0)]$certificate <#...#> )` are all possible to the user.

Reply:
I believe if we say about a script block we shouldn't concerned about .Net templates. Script writes don't see underlying implementations and usualy expect implicit `$PSItem` or explicit `Param`.
Also the variables actually is automatic variables and we should add its in list of automatic variables. This seems unjustified.
I suggest we wait for @lzybkr  opinion.

Reply:
It's automatic to the ScriptBlock only. These do not show up in the calling scope. I don't see the problem? Please elaborate "we should add its in list of automatic variables". Do Script Block have their own list of automatic variables?

And, we *should* be concerned about the underlying .NET. It would be confusing to maintainers if the `Func` delegate has one argument order but the script block has another. The user gets the convenience of the automatic variables or `Param()` if they don't like that `$args[0]` is the `HttpMessageRequest` . As I said, all 4 arguments are equally important. Some users may have rules which never even test the certificate (URL based acceptance).

Reply:
Why set the variables **and** pass them as args?

Also, why set `$ErrorActionPreference`? That might not be desirable.

Reply:
> Why set the variables and pass them as args?

I wanted to support parameter binding by type i.e. 

```powershell
param(
    [System.Security.Cryptography.X509Certificates.X509Certificate2]$cert,
# ...
````
I was trying to be flexible. 

> Also, why set `$ErrorActionPreference`? That might not be desirable.

The [planned] documented behavior is that that errors and exceptions result in certificate failures. I ran into some strange scenarios where if errors weren't terminating it would hang the thread. but, this was before I switched to a closure and before I was using `LanguagePrimitives.IsTrue()`. I'm not sure if it s still needed, I can check though.

Reply:
I'm definitely not in favor of both variables and parameters. Either way you need good documentation, but I prefer parameters.

And please do verify that setting `$ErrorActionPreference` is not needed.

Reply:
This is the draft documentation:
https://github.com/PowerShell/PowerShell-Docs/compare/staging...markekraus:CertificatevalidationScriptDraft 

I'm not sure I agree that parameters are preferable. Personally, I'm OK with either. But, I think most people would rather have auto-variables than have to declare  `param()` or try to remember which `$args` is which. in the end, this is a documentation heavy dependent feature no matter which way it is implemented.

Reply:
I'll let the @PowerShell/powershell-committee chime in on this one.

Reply:
Automatic variables removed per @PowerShell/powershell-committee decision. `$ErrorActionPreference` also appears to be no longer needed.